docs/advanced-features/sitemaps.md

```diff
@@ -0,0 +1,179 @@
+  const pages = await globby([
+    'pages/*.js', // Include all pages
+    '!pages/_*.js', // Exclude _app or _document
+    '!pages/api', // Exclude API Routes
+    '!pages/404.js', // Exclude 404
+  ])
+
+  const sitemap = `
+    <?xml version="1.0" encoding="UTF-8"?>
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+        ${pages
+          .map((page) => {
+            const path = page.replace('pages', '').replace('.js', '')
+            const route = path === '/index' ? '' : path
+
+            return `
+              <url>
+                  <loc>${`https://nextjs.org${route}`}</loc>
+              </url>
+            `
+          })
+          .join('')}
+    </urlset>
+    `
+
+  writeFileSync('public/sitemap.xml', sitemap)
+}
+
+generate()
+```
+
+> **Note:** This Node.js script is using [ES Modules](https://nodejs.org/api/esm.html), denoted by the `.mjs` extension. You could also write this script with [CommonJS](https://nodejs.org/api/modules.html#modules_modules_commonjs_modules) if you prefer.
+
+Finally, add `postbuild` script in your `package.json` to run this script after `next build` completes.
+Your generated sitemap file gets created at `public/sitemap.xml`, which is then served as a [static file](/docs/basic-features/static-file-serving.md) at the root of your site.
+
+```json
+{
+  "scripts": {
+    "build": "next build",
+    "postbuild": "node ./scripts/generate-sitemap.mjs",
+    "dev": "next dev",
+    "start": "next start",
+    "lint": "next lint"
+  }
+}
+```
+
+## External Content
+
+If you have externally hosted data (e.g., a CMS), you'll need to make an API request
+before you can create your sitemap. This implementation will vary depending on your data source,
+but the idea is similar. To demonstrate, I've created an example using placeholder data.
+
+First, create a new file at `pages/sitemap.xml.js`.
+
+```js
+import React from 'react'
+
+class Sitemap extends React.Component {
+  static async getInitialProps({ res }) {
+    const request = await fetch(EXTERNAL_DATA_URL)
+    const posts = await request.json()
+
+    res.setHeader('Content-Type', 'text/xml')
+    res.write(createSitemap(posts))
+    res.end()
+  }
+}
+
+export default Sitemap
+```
+
+When the route `/sitemap.xml` is initially loaded, we will fetch posts from an external data source
+and then write an XML file as the response.
+
+```js
+import React from 'react'
+
+const EXTERNAL_DATA_URL = 'https://jsonplaceholder.typicode.com/posts'
+
+const createSitemap = (posts) => `<?xml version="1.0" encoding="UTF-8"?>
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+        ${posts
+          .map(({ id }) => {
+            return `
+                <url>
+                    <loc>${`${EXTERNAL_DATA_URL}/${id}`}</loc>
+                </url>
+            `
+          })
+          .join('')}
+    </urlset>
+    `
+
+class Sitemap extends React.Component {
+  static async getInitialProps({ res }) {
```

Comment:
Same here, should we change `getInitialProps` to `getServerSideProps()`?

Reply:
https://github.com/vercel/next.js/pull/28346#discussion_r693091871